server/src/main/java/org/apache/cassandra/sidecar/acl/authentication/ReloadingJwtAuthenticationHandler.java

```diff
@@ -0,0 +1,179 @@
+                if (!roles.isEmpty())
+                {
+                    user.attributes().put(CASSANDRA_ROLES_ATTRIBUTE_NAME, String.join(CASSANDRA_ROLE_SPLITTER, roles));
+                }
+                handler.handle(Future.succeededFuture(user));
+            }
+            catch (Exception e)
+            {
+                LOGGER.debug("Error processing cassandra role from JWT token", e);
+                handler.handle(Future.failedFuture(new HttpException(UNAUTHORIZED.code(),
+                                                                     "Error processing cassandra role from token")));
+            }
+        });
+    }
+
+    /**
+     * {@link NoOpAuthenticationProvider} is used in {@link ReloadingJwtAuthenticationHandler}.
+     * {@link ReloadingJwtAuthenticationHandler} delegates authenticating user to delegate handler.
+     * Hence it uses no op authentication provider
+     */
+    protected static class NoOpAuthenticationProvider implements AuthenticationProvider
+    {
+        public static final NoOpAuthenticationProvider INSTANCE = new NoOpAuthenticationProvider();
+
+        private NoOpAuthenticationProvider()
+        {
+        }
+
+        @Override
+        public void authenticate(JsonObject credentials, Handler<AsyncResult<User>> resultHandler)
+        {
+            resultHandler.handle(Future.succeededFuture());
+        }
+    }
+
+    /**
+     * Periodic task to generate {@link OAuth2AuthHandlerImpl} with refreshed configuration from
+     * {@link OpenIDConnectAuth} discover.
+     */
+    private class OAuth2AuthHandlerGenerateTask implements PeriodicTask
+    {
+        public DurationSpec delay()
+        {
+            return jwtParameterExtractor.configDiscoverInterval();
+        }
+
+        public void execute(Promise<Void> promise)
+        {
+            OAuth2Options options = new OAuth2Options().setSite(jwtParameterExtractor.site())
+                                                       .setClientId(jwtParameterExtractor.clientId());
+
+            OpenIDConnectAuth.discover(vertx, options)
+                             .onSuccess(oAuthProvider -> {
+                                 OAuth2AuthHandlerImpl handler = new OAuth2AuthHandlerImpl(vertx, oAuthProvider, null);
+                                 if (!jwtParameterExtractor.scopes().isEmpty())
+                                 {
+                                     handler.withScopes(jwtParameterExtractor.scopes());
+                                 }
+                                 delegateHandler.set(handler);
+                                 promise.complete();
+                             })
+                             .onFailure(cause -> {
+                                 LOGGER.error("Error encountered during OpenID discovery", cause);
+                                 promise.fail(cause);
+                             });
+        }
+    }
+}
```

Comment:
I am a bit worried about the tight coupling between the refresh task and the authenticator. Also such periodic task is created per each `ReloadingJwtAuthenticationHandler` instance, which I do not think it is desired. Although `ReloadingJwtAuthenticationHandler` is effectively singleton, there is no strong guarantee on it. 

Can you move the `OAuth2AuthHandlerGenerateTask` to a dedicated class file and make it a singleton? It can register the "discover" listener/`ReloadingJwtAuthenticationHandler` and update the reference on success.

Reply:
I don't think `ReloadingJwtAuthenticationHandler` can be considered a singleton. You could configure multiple OpenID connect services in theory, and that would be valid. And you will need to independently refresh each of them

Reply:
Discussed offline, to have the task as it. But adding OpenID site and client id information to task name, to have 1 task per OpenID provider